content/PHPDataObjects/DataObjects.q.md

```diff
@@ -0,0 +1,1458 @@
+
+- It produces an error.
+
+- No output is displayed.
+
+- It prints `Unable to establish a connection.`.
+
+- It prints `Successfully connected to the database.` and `Unable to insert values into the table.`.
+
+- It prints `Successfully connected to the database.` and `Successfully inserted the values into the table.`.
+
+
+/// type=MS, answer=[4,5]
+
+Which statements correctly describe the error message?
+
+- On line 4, the `students` table does not exist.
+
+- There are no parentheses `()` that enclosed the SQL statement on line 4.
+
+- On line 4, the column names `first_name`, `last_name`, `birth_date`, and `gender` are not enclosed in single quotes `''`.
+
+- In `student.php`, there is no `VALUES` clause that specifies the values to be inserted into the selected columns on line 4.
+
+- On line 4, the statement `$sql = "INSERT INTO students (first_name, last_name, birth_date, gender)";` is incorrect.
+
+:::
+
+/// type=CR, answer=[tests/DataObjects/MissingValuesClause.php], filename=[connection.php,student.php]
+
+Correct the code so that it inserts the values `Samantha`, `Danes`, `1999-10-12`, and `Female` into the `students` table and displays the strings `Successfully connected to the database.` and `Successfully inserted the values into the table.`. 
+
+```php
+// connection.php
+<?php
+  $host = 'localhost';
+  $db = 'LibraryDB';
+  $port = '5432';
+  $username = 'postgres';
+  $password = 'Admin01';
+
+  $dsn = "pgsql:host=$host;port=$port;dbname=$db;user=$username;password=$password";
+
+  try {
+      $conn = new PDO($dsn);
+      if ($conn) {
+          echo "Successfully connected to the database. ";
+      }
+  } catch (Exception $e) {
+      echo "Unable to establish a connection."; 
+  }
+?>
+```
+
+```php
+// student.php
+<?php
+    require_once("connection.php");
+
+    $sql = "INSERT INTO students (first_name, last_name, birth_date, gender)";
+
+    try {
+        $stmt = $conn->query($sql);
+        if (!$stmt) {
+            throw new Exception('Unable to insert values into the table.');
+        }
+        echo "Successfully inserted the values into the table.";
+    } catch (Exception $e) {
+        echo $e->getMessage();
+    }
+?>
+```
+
+:::
+
+/// type=REPL, filename=[connection.php,student.php]
+
+```php
+// connection.php
+<?php
+  $host = 'localhost';
+  $db = 'LibraryDB';
+  $port = '5432';
+  $username = 'postgres';
+  $password = 'Admin01';
+
+  $dsn = "pgsql:host=$host;port=$port;dbname=$db;user=$username;password=$password";
+
+  try {
+      $conn = new PDO($dsn);
+      if ($conn) {
+          echo "Successfully connected to the database. ";
+      }
+  } catch (Exception $e) {
+      echo "Unable to establish a connection."; 
+  }
+?>
+```
+
+```php
+// student.php
+<?php
+    require_once("connection.php");
+
+    $stmt = $conn->query("SELECT * FROM students");
+    foreach ($stmt as $row) {
+        echo $row['student_id'].$row['first_name'].$row['last_name'].$row['birth_date'].$row['gender'].' ';
```

Comment:
I think this is still related to AR's comment about the whitespace at the end. Please check. Thank you.